### PR TITLE
executor: add retriver for tidb_hot_regions_history 

### PR DESCRIPTION
**Pull request by IcePigZDB**
Signed-off-by: IcePigZDB <icepigzdb@gmail.com>

Problem Summary: Create a new table `TIDB_HOT_REGIONS_HISTORY` in the `INFORMATION_SCHEMAT` database to retrieve history hotspot information stored by PD periodically.
### What problem does this PR solve?
Proposal: [pingcap/tidb#27487](https://github.com/pingcap/tidb/pull/27487)
Associated Issue: #25281

Associated PRs in TiDB:
PRs splited from oversize PR #27224 in merge order :
1. #27373 
2. #27374 
3. #27375  **<- current one** 
4. #27899
5. #27900   

Associated PR in PD:  
 PRs splited from old oversize one: tikv/pd/pull/3989 
* tikv/pd/pull/4019
* tikv/pd/pull/4020
* tikv/pd/pull/4021
### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
select * from information_schema.tidb_hot_regions_history where update_time>='2019-10-10 10:10:10' and update_time<'2019-10-11 10:10:10'
```

### CODE REVIEW

**Comment by IcePigZDB** [comment]
move test file of executor back to this PR, /close #27900 .

**Comment by IcePigZDB** [comment]
move test file of executor back to this PR, /close #27900 .

**Comment by nolouch** [comment]
@IcePigZDB  ci failed.

**Comment by nolouch** [comment]
PTAL @rleungx

File: executor/memtable_reader.go [hunk 1]
```diff
@@ -696,3 +701,299 @@ func (e *clusterLogRetriever) close() error {
...
+		}
+		// Initialize the heap
+		e.heap = &hotRegionsResponseHeap{}
+		for _, ch := range results {
+			result := <-ch
+			if result.err != nil {
+				sctx.GetSessionVars().StmtCtx.AppendWarning(result.err)
+			}
+			if result.err != nil || len(result.messages.HistoryHotRegion) == 0 {
+				continue
```
Comment (rleungx) [inline executor/memtable_reader.go:888]: move `continue` to line 886 and check `len(result.messages.HistoryHotRegion) > 0`?

File: executor/memtable_reader.go [hunk 2]
```diff
@@ -696,3 +701,299 @@ func (e *clusterLogRetriever) close() error {
...
+
+func (e *hotRegionsHistoryRetriver) getHotRegionRowWithSchemaInfo(
+	hisHotRegion *HistoryHotRegion,
+	tikvHelper *helper.Helper,
+	allSchemas []*model.DBInfo,
+	tz *time.Location,
+) ([]types.Datum, error) {
+	_, startKey, _ := codec.DecodeBytes(hisHotRegion.StartKey, []byte{})
+	_, endKey, _ := codec.DecodeBytes(hisHotRegion.EndKey, []byte{})
+	region := &tikv.KeyLocation{StartKey: startKey, EndKey: endKey}
```
Comment (rleungx) [inline executor/memtable_reader.go:934]: Is there possible the old table is dropped?

File: planner/core/memtable_predicate_extractor.go [hunk 3]
```diff
@@ -541,6 +541,24 @@ func (helper extractHelper) convertToTime(t int64) time.Time {
 	return time.Unix(0, t)
 }
 
+func (helper extractHelper) extractRoleBoolSliceFromUint64Slice(uint64Slice []uint64) []bool {
```
Comment (rleungx) [inline planner/core/memtable_predicate_extractor.go:544]: The function name is too long.

File: planner/core/util.go [hunk 4]
```diff
@@ -308,6 +308,19 @@ func extractStringFromUint64Slice(slice []uint64) string {
...
+// extractStringFromBoolSlice helps extract string info from bool slice.
+func extractStringFromBoolSlice(slice []bool) string {
+	if len(slice) < 1 {
+		return ""
+	}
+	l := make([]string, 0, len(slice))
+	for _, k := range slice {
+		l = append(l, fmt.Sprintf(`%t`, k))
+	}
+	sort.Strings(l)
```
Comment (rleungx) [inline planner/core/util.go:320]: What is the purpose of sort?

File: executor/memtable_reader.go (hunk 1 above)
Comment (IcePigZDB) [reply executor/memtable_reader.go:888]: done.

File: executor/memtable_reader.go (hunk 2 above)
Comment (IcePigZDB) [reply executor/memtable_reader.go:934]: yes, we have discussed this case, dropped tables' hot region will be ignore in current implement.
```go
	_, startKey, _ := codec.DecodeBytes(hisHotRegion.StartKey, []byte{})
	_, endKey, _ := codec.DecodeBytes(hisHotRegion.EndKey, []byte{})
	region := &tikv.KeyLocation{StartKey: startKey, EndKey: endKey}
	hotRange, err := helper.NewRegionFrameRange(region)
	if err != nil {
		return nil, err
	}

	f := tikvHelper.FindTableIndexOfRegion(allSchemas, hotRange)
	// Ignore row without corresponding schema f.
	if f == nil {
		return nil, nil
	}
```
This case has been tested in `memtable_reader_test.go/TestTiDBHotRegionsHistory` line 1097-1099, code run as expected.
```go
		//      table_id = 131, index_id = 1, index_value = 1, deleted schema
		{"2019-10-10 10:10:23", "UNKONW", "UNKONW", "131", "UNKONW", "1", "5", "5", "55555", "0", "1", "READ", "99", "99", "99", "99"},
		{"2019-10-10 10:10:24", "UNKONW", "UNKONW", "131", "UNKONW", "1", "6", "6", "66666", "0", "0", "WRITE", "99", "99", "99", "99"},
```

File: planner/core/memtable_predicate_extractor.go (hunk 3 above)
Comment (IcePigZDB) [reply planner/core/memtable_predicate_extractor.go:544]: change to `convertToBoolSlice` according to previous func `convertToTime`.

File: planner/core/util.go (hunk 4 above)
Comment (IcePigZDB) [reply planner/core/util.go:320]: In unnormal case, user may query `is_leader in(0,1)`, sort can keep the order of result string.

**Review by rleungx — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**